### PR TITLE
[miio] cloud login issues improvement

### LINKED ISSUE #7459: [miio] Error using Xiaomi cloud in miio binding
==> openhab.log <==
2020-04-24 08:54:45.437 [INFO ] [miio.internal.cloud.MiCloudConnector] - Error logging on to Xiaomi cloud (0): Error getting logon location URL. Return code: 70016

## Expected Behavior

Just found in logs. I filled creds to mihome, expected that binding will get tokens itself.

Would be great if it will fill tokens from cloud when i add thing

## Current Behavior

When i have humidifier thing in inbox, there is no token filled. Humidifier connected to mihome and credentials filled in binding

## Possible Solution
I'm a newbie, so no idea right now

## Steps to Reproduce (for Bugs)
- install miio binding
- fill creds to mihome (there should be something connected)
- wait for inbox
- add thing from inbox
- view config (should not be a token there)

## Your Environment

* Version used: openHAB 2.5.4 in docker using docker-compose
* Environment name and version (e.g. Chrome 76, Java 8, Node.js 12.9, ...): Synology 218+
* Operating System and version (desktop or mobile, Windows 10, Raspbian Buster, ...): DSM 6.2.2-24922 Update 6

### LINKED ISSUE #13611: [miio] Error loggin to Xiaomi cloud in miio binding -  Return code: 0
## Expected Behavior
Addon logs in to the Xiaomi cloud and retrieves the supported device list.

## Current Behavior
Login fails due to unexpected JSON response.

```
17:38:22.934 [TRACE] [.miio.internal.cloud.MiCloudConnector] - Xiaomi Login step 1
17:38:22.984 [TRACE] [.miio.internal.cloud.MiCloudConnector] - Xiaomi Login step 1 content response= &&&START&&&{"serviceParam":"{\"checkSafePhone\":false,\"checkSafeAddress\":false,\"lsrp_score\":0.0}","qs":"%3Fsid%3Dxiaomiio%26_json%3Dtrue","code":70016,"description":"登录验证失败","securityStatus":0,"_sign":"0psXfr43eNI0IX6q9Suk3qWbRqU=","sid":"xiaomiio","result":"error","captchaUrl":null,"callback":"https://sts.api.io.mi.com/sts","location":"https://account.xiaomi.com/fe/service/login?_json=true&sid=xiaomiio&qs=%253Fsid%253Dxiaomiio%2526_json%253Dtrue&callback=https%3A%2F%2Fsts.api.io.mi.com%2Fsts&_sign=0psXfr43eNI0IX6q9Suk3qWbRqU%3D&serviceParam=%7B%22checkSafePhone%22%3Afalse%2C%22checkSafeAddress%22%3Afalse%2C%22lsrp_score%22%3A0.0%7D&showActiveX=false&theme=&needTheme=false&bizDeviceType=","pwd":0,"child":0,"desc":"登录验证失败"}
17:38:22.985 [TRACE] [.miio.internal.cloud.MiCloudConnector] - Xiaomi Login step 1 response = HttpContentResponse[HTTP/1.1 200 OK - 760 bytes]
17:38:22.986 [TRACE] [.miio.internal.cloud.MiCloudConnector] - Xiaomi Login step 1 sign = 0psXfr43eNI0IX6q9Suk3qWbRqU=
17:38:22.986 [TRACE] [.miio.internal.cloud.MiCloudConnector] - Xiaomi Login step 2
17:38:23.047 [TRACE] [.miio.internal.cloud.MiCloudConnector] - Xiaomi login step 2 response = HttpContentResponse[HTTP/1.1 200 OK - 1014 bytes]
17:38:23.048 [TRACE] [.miio.internal.cloud.MiCloudConnector] - Xiaomi login step 2 content = &&&START&&&{"notificationUrl":"https://account.xiaomi.com/identity/authStart?sid=passport&context=sMlx600MPFo9smESNZmGDdkmc7e2Z4sYqW81z2t4F3LTh9vz7gf0sNeze4znW9ytwEsqYzgWj0KQ1dl2CDw9Sne6w9qMFRCC6zg0de7L_8y-8xeaHHo6nn35dLNtMQOi8NvYi-RGx0rkivMCpbaPnsFYzDoRliupXK-Ox_r5MWElldDjtKTCY2mxQGinj_eAQdcERP5iV2wG2TqxQdy3AdXAKmfyg6y044u1491wG7tosQAIQriMFmPnhZqBzQjNN8rBmHfnogcvbfxGV0-eqRJTN4XemWcLkO_lLNB0sC26tx4W1rQ2chwK1v5N3DOxJxGofmuXkmq1VqVBqFp7DdVWoy5nGsZnI3CK0WT9cfTQWCXpOeYyJRWGkLP35irZ_X4SukwbSFzQPgu5gMfMPkaSdY0ShZv-BACKoUYW_QQIrf1A0Rc0HReupBQNBbuDcQofNskAvVNLoUOcfV4MSb_s7PGsCh0UP4tk-LGenpNMhMhIT2tcEehGtG8gqgsqIFdYcK7Uwc5KmnGtdRvJTOhuvxUZXSuVB7-aoh86urjJl4n5VMMl7nWUXmZORNXYyfNZ4rhSOWzrBMZiTrQK0JewQkWO3YuXJrGe4h6x33rcSRHGwTPzo2cRpCMg_Xf5xTYe_Fqd-ggqp0NRzZEvntc7y_DRAb8PSo1k_RNf4yWYg7Vg-P4SwhhVUU2nUcQ7FFFJKwG5pooL7UZwbPI5PeU2Up4YEnmrQlQvcmxppnBJNj9PNIna0N8sQomhMl2E&_locale=de_DE","result":"ok","code":0,"captchaUrl":null,"description":"成功","location":"","securityStatus":16,"pwd":0,"child":0,"desc":"成功"}
17:38:23.049 [TRACE] [.miio.internal.cloud.MiCloudConnector] - Xiaomi login ssecurity =
17:38:23.050 [TRACE] [.miio.internal.cloud.MiCloudConnector] - Xiaomi login userId =
17:38:23.050 [TRACE] [.miio.internal.cloud.MiCloudConnector] - Xiaomi login cUserId =
17:38:23.051 [TRACE] [.miio.internal.cloud.MiCloudConnector] - Xiaomi login passToken =
17:38:23.052 [TRACE] [.miio.internal.cloud.MiCloudConnector] - Xiaomi login location =
17:38:23.052 [TRACE] [.miio.internal.cloud.MiCloudConnector] - Xiaomi login code = 0
17:38:23.053 [TRACE] [.miio.internal.cloud.MiCloudConnector] - Cookie dump for https://account.xiaomi.com/pass/serviceLoginAuth2
17:38:23.054 [TRACE] [.miio.internal.cloud.MiCloudConnector] - Cookie (.xiaomi.com) : sdkVersion --> accountsdk-18.8.15     (path: /. Removed: false)
17:38:23.055 [TRACE] [.miio.internal.cloud.MiCloudConnector] - Cookie (.xiaomi.com) : deviceId --> cuhdom     (path: /. Removed: false)
17:38:23.055 [TRACE] [.miio.internal.cloud.MiCloudConnector] - Cookie (.xiaomi.com) : uLocale --> de_DE     (path: /. Removed: false)
17:38:23.055 [TRACE] [.miio.internal.cloud.MiCloudConnector] - Cookie (account.xiaomi.com) : pass_ua --> web     (path: /. Removed: false)
17:38:23.056 [INFO ] [.miio.internal.cloud.MiCloudConnector] - Error logging on to Xiaomi cloud (0): Error getting logon location URL. Return code: 0
```

## Steps to Reproduce (for Bugs)

1. Start fresh openhab instance (3.3.0-alpine)
2. Install Miio binding addon
3. Enter Xiaomi cloud email and password to the addon configuration. Also set country to "cn" (I have a device that is supported only in cn region).
4. Try to add a new thing from Miio binding
5. Hit the "Scan" button
6. Check the openhab log. There will be login error.

## Context

It seems to be working from my home PC but not from cloud VPS. It looks like there is some kind of user verification step that is missing in openhab addon.

Unfortunately completing the verification flow using home PC browser doesn't help. Openhab addon is still unable to logon. It could be connected to an unverified IP address or something like that.

## Your Environment

* Version used: openhab:3.3.0-alpine build, Mi IO binding 3.3.0
* Operating System and version (desktop or mobile, Windows 10, Raspbian Buster, ...): Ubuntu, Oracle cloud VPS, ARM cpu

image

image

image

### PR DESCRIPTION
**Pull request by marcelrv**
* Small change in debug loggin is login issues are expected
* Additional trouble shooting steps in readme
* Fix duplicate name in device list preventing readme maker running
* Catch a json exception in readme maker

close #13611
close #7459

Signed-off-by: Marcel Verpaalen <marcel@verpaalen.com>

### CODE REVIEW

File: bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/cloud/CloudLogin1DTO.java [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+    public String getSid() {
+        return sid;
+    }
+
+    public String getResult() {
+        return result;
+    }
+
+    public Object getCaptchaUrl() {
```
Comment (lolodomo) [inline bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/cloud/CloudLogin1DTO.java:99]: Why returning Object and not String (while `captchaUrl` is of type String) ?

File: bundles/org.openhab.binding.miio/README.base.md [hunk 2]
```diff
@@ -172,7 +172,13 @@ Firmware of the device don't accept commands coming from other subnets.
 Set the communication in the thing configuration to 'cloud'.
 
 _Cloud connectivity is not working_
-The most common problem is a wrong userId/password. Try to fix your userId/password.
+The most common problem is a wrong userId/password. Try to fix your userId/password. The login userId & password are entered in the binding config page.
+If the problem persists you can try the following:
+
+* Your ip might need to be validated/confirmed. Logon to https://account.xiaomi.com/ **from the ip of your openHAB server** with a browser.
```
Comment (lolodomo) [inline bundles/org.openhab.binding.miio/README.base.md:178]: I do not understand the second sentence.

File: bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/cloud/CloudLogin1DTO.java (hunk 1 above)
Comment (marcelrv) [reply bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/cloud/CloudLogin1DTO.java:99]: I missed fixing than from the input I got from the json-to-pojo tool I used

File: bundles/org.openhab.binding.miio/README.base.md (hunk 2 above)
Comment (marcelrv) [reply bundles/org.openhab.binding.miio/README.base.md:178]: I rephrased it. Hope it is more clear this time

**Comment by marcelrv** [comment]
@lolodomo saw there was some merge conflict in the readme after merging of #13724
This is resolved now

File: bundles/org.openhab.binding.miio/README.base.md (hunk 2 above)
Comment (lolodomo) [reply bundles/org.openhab.binding.miio/README.base.md:178]: @jlaur : do you understand what means "Logon to https://account.xiaomi.com/ **from the ip of your openHAB server** with a browser."?

Comment (jlaur) [reply bundles/org.openhab.binding.miio/README.base.md:178]: I'm not sure I understand it, but let me explain how I understand it: Xiaomi need a kind of verification that a specific IP address is assigned to the user claiming it. Or, more specifically, the the IP address is currently assigned (dynamically or statically) by the ISP to the users router. So I assume we are talking about the public IP address. In this case I would assume that a browser on any computer within the LAN behind the router (assuming only one public IP address) can be used to perform this "validation". So I don't understand it like I would need to install and try to run some browser on my openHABian installation running on a Raspberry Pi.

In other words, back to my intro: No, I'm not sure I understand it, I would need some more context, and I think it could be further rewritten after that.

Comment (lolodomo) [reply bundles/org.openhab.binding.miio/README.base.md:178]: @marcelrv : we have clearly understood that. What is not clear for us is the new instruction you are providing in the documentation: "Logon to https://account.xiaomi.com/ from the ip of your openHAB server with a browser.".
If it is not clear for both of us, I am afraid it will not be clear for many users. Only this sentence has to be clarified to be understandable what exactly has to be done.
I am sorry in case it is obvious for you.

Comment (marcelrv) [reply bundles/org.openhab.binding.miio/README.base.md:178]: _If it is not clear for both of us, I am afraid it will not be clear for many users._ fully agree. 
the matter indeed is how to best describe this so it is clear to everyone

Comment (lolodomo) [reply bundles/org.openhab.binding.miio/README.base.md:178]: Maybe a screen capture could help?

Comment (marcelrv) [reply bundles/org.openhab.binding.miio/README.base.md:178]: I'll check with @xduseko if I can use his screenshot. I can't reproduce this myself

File: bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/cloud/MiCloudConnector.java [hunk 3]
```diff
@@ -411,8 +411,9 @@ private String loginStep1() throws InterruptedException, TimeoutException, Execu
         logger.trace("Xiaomi Login step 1 response = {}", responseStep1);
         try {
             JsonElement resp = JsonParser.parseString(parseJson(content));
-            if (resp.isJsonObject() && resp.getAsJsonObject().has("_sign")) {
-                String sign = resp.getAsJsonObject().get("_sign").getAsString();
+            CloudLogin1DTO jsonResp = GSON.fromJson(resp, CloudLogin1DTO.class);
+            final String sign = jsonResp.getSign();
+            if (sign != null && !jsonResp.getSign().isBlank()) {
```
Comment (lolodomo) [inline bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/cloud/MiCloudConnector.java:416]: `if (sign != null && !sign.isBlank()) {`

**Review by lolodomo — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.binding.miio/README.base.md (hunk 2 above)
Comment (lolodomo) [reply bundles/org.openhab.binding.miio/README.base.md]: The sentence I did not understand is still there, unchanged ; I am not sure the added screen capture is really helpful to understand it.
But ok, I will not insist more, let's assume the users will understand better than me what you mean.

Comment (marcelrv) [reply bundles/org.openhab.binding.miio/README.base.md]: I did one more attempt to clarify it better, and trying to understand in which scenario's there might be issues.
The vast majority won't face this as their (external) ip is equal to their regular  pc/browser ip.

File: bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/cloud/MiCloudConnector.java (hunk 3 above)
Comment (marcelrv) [reply bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/cloud/MiCloudConnector.java]: thanks. fixed

File: bundles/org.openhab.binding.miio/README.base.md (hunk 2 above)
Comment (lolodomo) [reply bundles/org.openhab.binding.miio/README.base.md]: Ok thank you, I finally understood.
Maybe you could replace "connect **from the ip of your openHAB server** with a browser" by "connect with your browser **while no VPN is enabled** (to not hide your external IP address)". WDYT ?

And the screen capture is no more necessary. WDYT ?

Comment (marcelrv) [reply bundles/org.openhab.binding.miio/README.base.md]: Maybe you could replace "connect from the ip of your openHAB server with a browser" by "connect with your browser while no VPN is enabled (to not hide your external IP address)". WDYT ?
I think that is the most likely usecase indeed, but I think the user creating the issue actually was hosting his OH server somewhere else (on a VPS), why I understood he made a bridge to that server and was running a browser from the ip of his hosting provider. (He was by running some vpn docker with port forwarding, he also could have used some SSH tunneling I think) 
this is why I was emphasizing the ' from the ip of your openHAB server ' part... 

Do you think it is more clear if i change it to something like "browse with the same external ip address as your openHAB server to...http etc"

The screenshot does not contribute much, no problem removing it...

Comment (lolodomo) [reply bundles/org.openhab.binding.miio/README.base.md]: > Do you think it is more clear if i change it to something like "browse with the same external ip address as your openHAB server to...http etc"

Yes, I believe

**Review by lolodomo — APPROVED** [review]
LGTM, thank you
